libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt

```diff
@@ -599,6 +615,146 @@ class MapboxNavigationTest {
+        val navigationRerouteController: NavigationRerouteController? = mockk(relaxed = true) {
+            every { reroute(any<NavigationRerouteController.RoutesCallback>()) } answers {
+                (firstArg() as NavigationRerouteController.RoutesCallback)
+                    .onNewRoutes(newRoutes, mockk())
+            }
+        }
+        coEvery {
+            tripSession.setRoutes(any(), any(), any())
+        } returns NativeSetRouteValue(emptyList())
+
+        createMapboxNavigation()
+        mapboxNavigation.setRerouteController(navigationRerouteController)
+        val observers = mutableListOf<OffRouteObserver>()
+        verify { tripSession.registerOffRouteObserver(capture(observers)) }
+
+        observers.forEach {
+            it.onOffRouteStateChanged(true)
+        }
+        coVerify(exactly = 1) {
+            directionsSession.setRoutes(newRoutes, 0, RoutesExtra.ROUTES_UPDATE_REASON_REROUTE)
+        }
+    }
+
+    @Test
+    fun `new routes are not set after reroute if they are invalid`() {
+        val newRoutes = listOf(mockk<NavigationRoute>(), mockk())
+        val navigationRerouteController: NavigationRerouteController? = mockk(relaxed = true) {
+            every { reroute(any<NavigationRerouteController.RoutesCallback>()) } answers {
+                (firstArg() as NavigationRerouteController.RoutesCallback)
+                    .onNewRoutes(newRoutes, mockk())
+            }
+        }
+        coEvery {
+            tripSession.setRoutes(any(), any(), any())
+        } returns NativeSetRouteError("some error")
+
+        createMapboxNavigation()
+        mapboxNavigation.setRerouteController(navigationRerouteController)
+        val observers = mutableListOf<OffRouteObserver>()
+        verify { tripSession.registerOffRouteObserver(capture(observers)) }
+
+        observers.forEach {
+            it.onOffRouteStateChanged(true)
+        }
+        coVerify(exactly = 0) {
+            directionsSession.setRoutes(any(), any(), any())
+        }
+    }
+
+    @Test
+    fun `set reroute controller in fetching state sets routes to session`() {
+        val newRoutes = listOf(mockk<NavigationRoute>(), mockk())
+        val navigationRerouteController: NavigationRerouteController? = mockk(relaxed = true) {
+            every { state } returns RerouteState.FetchingRoute
+            every { reroute(any<NavigationRerouteController.RoutesCallback>()) } answers {
+                (firstArg() as NavigationRerouteController.RoutesCallback)
+                    .onNewRoutes(newRoutes, mockk())
+            }
+        }
+        coEvery {
+            tripSession.setRoutes(any(), any(), any())
+        } returns NativeSetRouteValue(emptyList())
+
+        createMapboxNavigation()
+        mapboxNavigation.setRerouteController(navigationRerouteController)
+        coVerify(exactly = 1) {
+            directionsSession.setRoutes(newRoutes, 0, RoutesExtra.ROUTES_UPDATE_REASON_REROUTE)
```

Comment:
I'm wondering, do we need to check other side effects of set routes like `routeAlternativesController.processAlternativesMetadata`? 
@dzinad , @RingerJK , what do you think?

Reply:
I'd say yes. However, here my purpose was to cover all the entry points for `internalSetNavigationRoutes` so that I can check the behaviour depending on whether it's success of failure. I think that goal was achieved. I don't see why not add this tests, but it's not really relevant to the task. I prefer doing such things separately so that the original issue does not get stuck because of these small improvements. I can cut a ticket if we agree to add the tests.

Reply:
`routeAlternativesController.processAlternativesMetadata` doesn't affect reason or otherwise, if you mean it